aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,742 @@
+            if not b['time_completed'] or not b['complete']:
+                skipped += 1
+                continue
+
+            time_completed = parse_hail_time(b['time_completed'])
+            in_date_range = start <= time_completed < end
+
+            if time_completed < start:
+                logger.info(
+                    f'{billing_project} :: Got {len(batches)} batches '
+                    f'in {n_requests} requests, skipping {skipped}'
+                )
+                return batches
+            if in_date_range:
+                batches.append(b)
+            else:
+                skipped += 1
+
+
+async def get_jobs_for_batch(batch_id, token: str) -> List[str]:
+    """
+    For a single batch, fill in the 'jobs' field.
+    """
+    jobs = []
+    last_job_id = None
+    end = False
+    iterations = 0
+
+    async with aiohttp.ClientSession() as session:
+        while not end:
+            iterations += 1
+
+            if iterations > 1 and iterations % 5 == 0:
+                logger.info(f'On {iterations} iteration to load jobs for {batch_id}')
+
+            q = '?limit=9999'
+            if last_job_id:
+                q += f'&last_job_id={last_job_id}'
+            url = HAIL_JOBS_API.format(batch_id=batch_id) + q
+
+            jresponse = await async_retry_transient_get_json_request(
+                url,
+                aiohttp.ClientError,
+                session=session,
+                headers={'Authorization': 'Bearer ' + token},
+            )
+            new_last_job_id = jresponse.get('last_job_id')
+            if new_last_job_id is None:
+                end = True
+            elif last_job_id and new_last_job_id <= last_job_id:
+                raise ValueError('Something fishy with last job id')
+            last_job_id = new_last_job_id
+            jobs.extend(jresponse['jobs'])
+
+    return jobs
+
+
+RE_matcher = re.compile(r'-\d+$')
+
+
+def billing_row_to_topic(row, dataset_to_gcp_map) -> Optional[str]:
+    """Convert a billing row to a dataset"""
+    project_id = row['project']['id']
+    topic = dataset_to_gcp_map.get(project_id, project_id)
+    if not topic:
+        return 'admin'
+
+    topic = RE_matcher.sub('', topic)
+    return topic
+
+
+def insert_new_rows_in_table(
+    table: str, obj: List[Dict[str, Any]], dry_run: bool
+) -> int:
+    """Insert JSON rows into a BQ table"""
+
+    if not obj:
+        logger.info('Not inserting any rows')
+        return 0
+
+    _query = f"""
+        SELECT id FROM `{table}`
+        WHERE id IN UNNEST(@ids);
+    """
+
+    ids = set(o['id'] for o in obj)
+    if len(ids) != len(obj):
+        counter = defaultdict(int)
+        for o in obj:
+            counter[o['id']] += 1
+        duplicates = [f'{k} ({v})' for k, v in counter.items() if v > 1]
+        raise ValueError(
+            'There are multiple rows with the same id: ' + ', '.join(duplicates)
+        )
+
+    job_config = bq.QueryJobConfig(
```

Comment:
Is there a maximum size here that's practical? Should we consider batching if there are too many elements, eg: 10,000?

Reply:
BQ says "10MB" and max param size of 10,000, so we should definitely consider and handle both of these cases in this function

Reply:
Rolled a max 9000 rows to insert + checks for max entry size (though should never get close).